edit_questionpy_form.php

```diff
@@ -154,19 +273,35 @@ public function set_data($question): void {
     }
 
     /**
-     * Load in existing data as form defaults. Usually new entry defaults are stored directly in
-     * form definition (new entry form); this function is used to load in data where values
-     * already exist and data is being edited (edit entry form).
-     *
-     * note: $slashed param removed
+     * Validates selected or uploaded package.
      *
      * @param array $data
      * @param array $files
      * @return array $errors
+     * @throws moodle_exception
      */
     public function validation($data, $files) {
+        global $USER;
         $errors = parent::validation($data, $files);
-        // TODO.
+
+        $source = $data['qpy_package_source'] ?? null;
+        if ($source == 'search') {
+            if (empty($data['qpy_package_hash'])) {
+                $errors['qpy_package_container'] = get_string('required');
+            }
+        } else if ($source == 'upload') {
+            $filestorage = get_file_storage();
+            if (isset($data['qpy_package_path_name_hash'])) {
+                if (!$filestorage->file_exists_by_hash($data['qpy_package_path_name_hash'])) {
+                    $errors['qpy_package_file'] = get_string('required');
+                }
+            } else {
+                $usercontext = context_user::instance($USER->id);
+                if (!$filestorage->get_area_files($usercontext->id, 'user', 'draft', $data['qpy_package_file'])) {
+                    $errors['qpy_package_file'] = get_string('required');
+                }
+            }
+        }
```

Comment:
Das scheint nicht so sauber zu funktionieren. Wenn ich eine neue Frage erstellen will und kein Paket auswähle (1) bzw. keines Hochlade (2), erhalte ich unschöne Fehlermeldungen statt den hinweis, dass da was required ist.

(1): `Exception - qtype_questionpy\package\package::get_by_version(): Argument #1 ($pkgversionid) must be of type int, null given, called in [dirroot]/question/type/questionpy/edit_questionpy_form.php on line 191`

(2): `Coding error detected, it must be fixed by a programmer: draft file with id '298790094' does not exist`